src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentDocumentClassifierPass.cs

```diff
@@ -0,0 +1,99 @@
+using System;
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Blazor.Razor
+{
+    internal class ComponentDocumentClassifierPass : DocumentClassifierPassBase, IRazorDocumentClassifierPass
+    {
+        protected override string DocumentKind => "Blazor.Component-0.1";
+
+        protected override bool IsMatch(RazorCodeDocument codeDocument, DocumentIntermediateNode documentNode)
+        {
+            // Treat everything as a component by default if Blazor is part of the configuration.
+            return true;
+        }
+
+        protected override void OnDocumentStructureCreated(
+            RazorCodeDocument codeDocument, 
+            NamespaceDeclarationIntermediateNode @namespace, 
+            ClassDeclarationIntermediateNode @class, 
+            MethodDeclarationIntermediateNode method)
+        {
+            @namespace.Content = (string)codeDocument.Items[BlazorCodeDocItems.Namespace];
+            if (@namespace.Content == null)
+            {
+                @namespace.Content = "Blazor";
+            }
+
+            @class.BaseType = BlazorApi.BlazorComponent.FullTypeName;
+            @class.ClassName = (string)codeDocument.Items[BlazorCodeDocItems.ClassName];
+            if (@class.ClassName == null)
+            {
+                @class.ClassName = codeDocument.Source.FilePath == null ? null : Path.GetFileNameWithoutExtension(codeDocument.Source.FilePath);
+            }
+
+            if (@class.ClassName == null)
+            {
+                @class.ClassName = "__BlazorComponent";
+            }
+
+            @class.Modifiers.Clear();
+            @class.Modifiers.Add("public");
+
+            // Note that DefaultDocumentWriter's VisitMethodDeclaration in 2.0 is hardcoded to
+            // emit methods with no parameters.
+            //
+            // This was added in 2.1, but we still need a workaround to support 2.0. In that case we
+            // inject the parameter later in RazorCompiler.
+            method.ReturnType = "void";
+            method.MethodName = BlazorApi.BlazorComponent.BuildRenderTree;
+            method.Modifiers.Clear();
+            method.Modifiers.Add("protected");
+            method.Modifiers.Add("override");
+
+            method.Parameters.Clear();
+            method.Parameters.Add(new MethodParameter()
+            {
+                ParameterName = "builder",
+                TypeName = BlazorApi.RenderTreeBuilder.FullTypeName,
+            });
+
+            // We need to call the 'base' method as the first statement.
+            var callBase = new CSharpCodeIntermediateNode();
+            callBase.Children.Add(new IntermediateToken
+            {
+                Kind = TokenKind.CSharp,
+                Content = $"base.{BlazorApi.BlazorComponent.BuildRenderTree}(builder);" + Environment.NewLine
+            });
+            method.Children.Insert(0, callBase);
+        }
+
+        #region Workaround
```

Comment:
This is simplest, most isolated workaround I could find for not being able to replace the intermediate node writer.